### ISSUE TITLE
TYPE_CHECKING block fooled by typing alias

### ISSUE THREAD

**Issue by mscuthbert** [astral-sh/ty#3779]
### Summary

TYPE_CHECKING blocks are not discovered when using a qualified name to evaluate.  The type narrowing from `int | None` to just `int` is properly discovered on `if TYPE_CHECKING` but not `if typing.TYPE_CHECKING` or `import typing as t; ... if t.TYPE_CHECKING`.  

In case it's helpful, this corresponds to pylint issue https://github.com/pylint-dev/pylint/issues/7524 - fixed in two stages there: https://github.com/pylint-dev/pylint/pull/6787/changes  https://github.com/pylint-dev/pylint/pull/7525/changes  -- the real world example is https://github.com/cuthbertlab/music21 (2.5k stars) where each module calls `import typing as t` and uses `t.XYZ` for all type functions.

`main.py`:

```py
from random import choice

from typing import TYPE_CHECKING
import typing
import typing as t

x: dict[str, int|None] = {'val': 1, 'bad': None}

# this hack always spells 'val' but coerces making it "val" 
# to fool ty into not knowing that the answer is Literal[1]... (TOO SMART!)
v_a_l = 'v' + choice(['a', 'a']) + 'l'

# now a is int|None, type checker can't know.  That's fine.
a = x[v_a_l]
if TYPE_CHECKING:
    assert isinstance(a, int)
a2 = a + 1

b = x[v_a_l]
if typing.TYPE_CHECKING:
    assert isinstance(b, int)
b2 = b + 1  # fails: b2 Unknown, Operator + not supported

c = x[v_a_l]
if t.TYPE_CHECKING:
    assert isinstance(c, int)
c2 = c + 1  # fails: c2 Unknown, Operator + not supported

```

`ty.json`:

```json
{
    "environment": {
        "python-version": "3.14"
    },
    "rules": {
        "undefined-reveal": "ignore"
    }
}
```

[Playground](https://play.ty.dev/dc621b2b-e76f-4c61-9238-25fbf77402da)

Thanks!

### Version

ty 0.0.49 (b669da87d 2026-06-11)

**Comment by mscuthbert** [comment]
Looking at https://github.com/astral-sh/ruff/blob/main/crates/ruff_python_semantic/src/analyze/typing.rs#L419 -- it appears the team has already implemented aliases in `pub fn is_type_checking_block(...)` -- similar to the pylint detectors (even better perhaps).  `is_type_checking_condition` should work for `typing.TYPE_CHECKING` at least if not for the `t.TYPE_CHECKING` condition, so it may be somewhere else.

**Comment by AlexWaygood** [comment]
The `ruff_python_semantic` crate is specific to the Ruff linter; it's not used by ty at all. For ty, the implementation will be in `ty_python_semantic` and/or `ty_python_core`

**Comment by AlexWaygood** [comment]
We appear to resolve the types of `typing.TYPE_CHECKING` and `t.TYPE_CHECKING` as `Literal[True] (Final)`, the same as we do for `TYPE_CHECKING`, which is interesting...

**Comment by AlexWaygood** [comment]
FWIW, doing your type narrowing in an `if TYPE_CHECKING` block and then using the narrowed type after the `if TYPE_CHECKING` block ends is quite unsound, so I'm not sure I endorse what you're doing here 😉 but I agree that `if TYPE_CHECKING` should behave the same way as `if typing.TYPE_CHECKING` here, and this is easy to fix.

**Comment by mscuthbert** [comment]
> FWIW, doing your type narrowing in an `if TYPE_CHECKING` block and then using the narrowed type after the `if TYPE_CHECKING` block ends is quite unsound, so I'm not sure I endorse what you're doing here 😉

Thanks and yes, this is probably the kick in the pants I need to migrate 99% of these to `casts` or `TypeGuards`, but definitely appreciate the quick PR love!  And looking forward to using ty in my own projects very soon!
